File: python/linked_list/copy_random_list.py
from typing import Optional

class Node:
    def __init__(self, val: int, next: 'Node' = None, random: 'Node' = None):
        self.val = val
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: Optional[Node]) -> Optional[Node]:
        if not head:
            return None

        current = head

        while current:
            new_node = Node(current.val, current.next, None)

            current.next = new_node

            current = new_node.next


        current = head

        while current:
            if current.random:
                current.next.random = current.random.next

            current = current.next.next

        original = head
        copy_head = head.next
        copy = copy_head

        while original:
            original.next = original.next.next

            if copy.next:
                copy.next = copy.next.next

            original = original.next

            copy = copy.next

        return copy_head

File: golang/linked_list/copy_random_list.go
package main

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}

	current := head

	for current != nil {
		newNode := &Node{Val: current.Val, Next: current.Next}

		current.Next = newNode

		current = newNode.Next
	}

	current = head

	for current != nil {
		if current.Random != nil {
			current.Next.Random = current.Random.Next
		}

		current = current.Next.Next
	}

	original := head
	copyHead := head.Next

	copy := copyHead

	for original != nil {
		original.Next = original.Next.Next

		if copy.Next != nil {
			copy.Next = copy.Next.Next
		}

		original = original.Next

		copy = copy.Next
	}

	return copyHead
}
